File: ShortUrl.js
var util = require('util');
var DEFAULT_ALPHABET = 'mn6j2c4rv8bpygw95z7hsdaetxuk3fq'
var DEFAULT_BLOCK_SIZE = 24
var MIN_LENGTH = 6

function ShortEncoder (alphabet, block_size) {
	if(!alphabet){
		this.alphabet = DEFAULT_ALPHABET;
	}
	if(!block_size){
		this.block_size = DEFAULT_BLOCK_SIZE;
	}
	this.mask = (1 << this.block_size) - 1;

	var tmpMapping = [];
	for(var i=this.block_size-1;i>=0;i--){
		tmpMapping.push(i);
	}
    this.mapping = tmpMapping;
}


ShortEncoder.prototype.encode = function(n){
    return (n & ~this.mask) | this._encode(n & this.mask);
};

ShortEncoder.prototype._encode = function(n){
    var result = 0
    for(var i in this.mapping){
    	var b = this.mapping[i];

    	if (n & (1 << i))
            result |= (1 << b);
    }

    return result;
};

ShortEncoder.prototype.decode = function(n){
    return (n & ~this.mask) | this._decode(n & this.mask);
};

ShortEncoder.prototype._decode = function(n){
    var result = 0
    for(var i in this.mapping){
    	var b = this.mapping[i];

    	if (n & (1 << b))
            result |= (1 << i);
    }
        
    return result;
};

ShortEncoder.prototype.enbase = function(x, min_length){
	min_length = min_length || MIN_LENGTH;

    var result = this._enbase(x);
    var padding = '';
    var paddingTimes = (min_length - result.length);
    if(paddingTimes > 0){
    	var tmp = [];
    	for(var i=0;i<paddingTimes;i++){
    		tmp.push(this.alphabet[0]);
    	}

    	padding = tmp.join('');
    }
    
    return util.format('%s%s', padding, result);
};

ShortEncoder.prototype._enbase = function(x){
    var n = this.alphabet.length;
    if (x < n)
        return this.alphabet[x];

    return this._enbase((x / n) | 0) + this.alphabet[(x % n) | 0];
};

ShortEncoder.prototype.debase = function(x){
    var n = this.alphabet.length;
    var result = 0;
    var tmp = x.split('').reverse().join('');
    for(var i in tmp){
    	var c = tmp[i];
    	result += this.alphabet.indexOf(c) * (Math.pow(n, i));
    }
        
    return result;
};

/**
 * shorten encode
 * @param  {int} n 
 * @return {string}   
 */
ShortEncoder.prototype.shorten = function (n) {
	return this.enbase(this.encode(n));
};

// var shortEncoder = new ShortEncoder();

// for(var i=1;i<10;i++){
// 	var a = shortEncoder.encode(i);
// 	var b = shortEncoder.enbase(a);
// 	var c = shortEncoder.debase(b);
// 	var d = shortEncoder.decode(c);

// 	console.log('%s %s %s %s %s', i, a, b, shortEncoder.shorten(i), c, d);
// }

File: ShortUrl.cs
public class ShortUrl
{
    const string DEFAULT_ALPHABET = "mn6j2c4rv8bpygw95z7hsdaetxuk3fq";
    const int DEFAULT_BLOCK_SIZE = 24;
    const int MIN_LENGTH = 5;

    public ShortUrl() 
    {
        this.Alphabet = DEFAULT_ALPHABET;
        this.BlockSize = DEFAULT_BLOCK_SIZE;
        this.Mark = (1 << DEFAULT_BLOCK_SIZE) - 1;

        int[] array = new int[DEFAULT_BLOCK_SIZE];
        for (int i = 0; i < DEFAULT_BLOCK_SIZE; i++)
        {
            array[i] = i;
        }

        this.Mapping = array;
        this.Mapping = this.Mapping.Reverse<int>().ToArray<int>();    
    }

    private string Alphabet
    {
        get;
        set;
    }

    private int BlockSize
    {
        get;
        set;
    }

    private int Mark
    {
        get;
        set;
    }

    private int[] Mapping
    {
        get;
        set;
    }

    public int Encode(int n)
    {
        return (n & ~this.Mark) | this._encode(n & this.Mark);
    }
    
    private int _encode(int n)
    {
        int result = 0;
        for(int i=0;i<this.Mapping.Length;i++)
        {
            int b = this.Mapping[i];
            if ((n & (1 << i)) > 0)
                result |= (1 << b);
        }
        
        return result;
    }
    
    public string Enbase(int x, int min_length = MIN_LENGTH)
    {
        string result = this._enbase(x);
        int nP = min_length - result.Length;
        string padding = string.Empty;
        if (nP > 0)
        {
            padding = padding.PadLeft(nP, this.Alphabet[0]);
        }

        return padding + result;
    }

    private string _enbase(int x)
    {
        int n = this.Alphabet.Length;
        if (x < n)
            return this.Alphabet[x].ToString();

        return this._enbase(x / n) + this.Alphabet[x % n].ToString();
    }

    public int debase(string x)
    {
        int n = this.Alphabet.Length;
        int result = 0;
        char[] array = x.Reverse<char>().ToArray<char>();

        for(int i=0;i<array.Length;i++)
        {
            char c = array[i];
            result += this.Alphabet.IndexOf(c) * Convert.ToInt32(Math.Pow(n, i));
        }

        return result;
    }

    public int decode(int n)
    {
        return (n & ~this.Mark) | this._decode(n & this.Mark);
    }

    private int _decode(int n)
    {
        int result = 0;
        for(int i=0;i<this.Mapping.Length;i++)
        {
            int b = this.Mapping[i];

            if ((n & (1 << b)) > 0)
                result |= (1 << i);
        }

        return result;
    }
}